src/platforms/posix/main.cpp

```diff
@@ -32,134 +32,236 @@
-	void _SigIntHandler(int sig_num)
-	{
-		cout.flush();
-		cout << endl << "Exiting..." << endl;
-		cout.flush();
-		_ExitFlag = true;
+	if (strstr(argv[0], prefix)) {
+		is_client = true;
 	}
-	void _SigFpeHandler(int sig_num);
-	void _SigFpeHandler(int sig_num)
-	{
-		cout.flush();
-		cout << endl << "floating point exception" << endl;
-		PX4_BACKTRACE();
-		cout.flush();
+
+	if (argc < 2) {
+		PX4_ERR("Not enough arguments.");
+		print_usage();
+		return -1;
 	}
-}
 
-static void print_prompt()
-{
-	cout.flush();
-	cout << "pxh> ";
-	cout.flush();
-}
+	if (is_client) {
 
-static void run_cmd(const vector<string> &appargs, bool exit_on_fail, bool silently_fail = false)
-{
-	// command is appargs[0]
-	string command = appargs[0];
+		if (!is_already_running()) {
+			PX4_ERR("PX4 daemon not running yet");
+			return -1;
+		}
 
-	if (apps.find(command) != apps.end()) {
-		const char *arg[appargs.size() + 2];
+		/* Remove the prefix by moving argv[0] and 1 more for the 0 termination. */
+		memmove(argv[0], argv[0] + strlen(prefix), strlen(argv[0]) + 1);
 
-		unsigned int i = 0;
+		px4_daemon::Client client;
+		client.generate_uuid();
+		client.register_sig_handler();
+		return client.process_args(argc, (const char **)argv);
 
-		while (i < appargs.size() && appargs[i] != "") {
-			arg[i] = (char *)appargs[i].c_str();
-			++i;
-		}
+	} else {
+		/* Server/daemon apps need to parse the command line arguments. */
+		for (int i = 1; i < argc; ++i) {
+			if (argv[i][0] == '-') {
+
+				if (strcmp(argv[i], "-h") == 0) {
+					print_usage();
+					return 0;
+
+				} else if (strcmp(argv[i], "-d") == 0) {
+					pxh_off = true;
 
-		arg[i] = (char *)0;
+				} else if (strcmp(argv[i], "-c") == 0) {
+					chroot_on = true;
 
-		int retval = apps[command](i, (char **)arg);
+				} else {
+					PX4_ERR("Unknown/unhandled parameter: %s", argv[i]);
+					print_usage();
+					return 1;
+				}
+
+			} else {
+				// This is an argument that does not have '-' it must be the file name.
+				std::ifstream infile(argv[i]);
 
-		if (retval) {
-			cout << "Command '" << command << "' failed, returned " << retval << endl;
+				if (infile.good()) {
+					infile.close();
+					commands_file = argv[i];
 
-			if (exit_on_fail && retval) {
-				exit(retval);
+				} else {
+					PX4_ERR("Error opening file: %s", argv[i]);
+					return -1;
+				}
 			}
 		}
 
-	} else if (command.compare("help") == 0) {
-		list_builtins();
+		if (is_already_running()) {
+			PX4_ERR("PX4 daemon already running");
+			return -1;
+		}
+
+		if (chroot_on) {
+			// Lock this application in the current working dir
+			// this is not an attempt to secure the environment,
+			// rather, to replicate a deployed file system.
+			use_chroot();
+		}
+
+		px4_daemon::Server server;
+		server.start();
+
+		register_sig_handler();
+		set_cpu_scaling();
+
+		DriverFramework::Framework::initialize();
+
+		px4::init_once();
+		px4::init(argc, argv, "px4");
+
+		run_startup_bash_script(commands_file);
```

Comment:
This does not work with chroot for 2 reasons:
- bash will not be found when chrooting
- the commands file is opened for test before chrooting and then afterwards again. one of them will probably fail, since chroot changes the root.

do we even need chroot?